HLTrigger/special/plugins/HLTPixelTrackFilter.cc

```diff
@@ -60,12 +63,22 @@ void HLTPixelTrackFilter::fillDescriptions(edm::ConfigurationDescriptions& descr
 // member functions
 //
 // ------------ method called to produce the data  ------------
-bool HLTPixelTrackFilter::filter(edm::Event& event, const edm::EventSetup& iSetup) {
+bool HLTPixelTrackFilter::hltFilter(edm::Event& event,
+                                    const edm::EventSetup& iSetup,
+                                    trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+  if (saveTags())
+    filterproduct.addCollectionTag(inputTag_);
+
   // get hold of products from Event
-  edm::Handle<reco::TrackCollection> trackColl;
-  event.getByToken(inputToken_, trackColl);
+  const auto& trackColl = event.getHandle(inputToken_);
 
   unsigned int numTracks = trackColl->size();
+  for (size_t i = 0; i < numTracks; i++)
+    filterproduct.addObject(trigger::TriggerTrack, reco::RecoChargedCandidateRef(trackColl, i));
```

Comment:
what's the expected increase of event size?

Reply:
This is meant to be run in ultra-peripheral collisions, in the low pT pixel track triggers that has a maximum of 400 pixel clusters (https://github.com/cms-sw/cmssw/blob/CMSSW_13_2_X/HLTrigger/Configuration/python/HLT_HIon_cff.py#L25389) where there are most of the time <10 tracks per event (mostly the events are either empty or have around 2 tracks for UPC coherent photo-production processes). I will try to derive the increase in event size by re-emulating on a 2023 PbPb HIForward dataset.

Reply:
> This is meant to be run in PbPb ultra-peripheral collisions

Since someone in the future may decide to use this filter on a completely different set of events, would it make sense to add a flag to disable this feature ? (e.g. a boolean `fillTriggerObjectWithRefs`)

Reply:
I agree that this opening a dangerous door. Please add a parameter to enable that feature (and leave it disable by default).

Reply:
Sorry, no additional parameter; this is what ```saveTags = cms.bool( True ),``` (or ```False```) is for (parameter defined in the base class ```HLTFilter```).

Reply:
sorry, see https://github.com/cms-sw/cmssw/pull/45373/files#r1666449069

Reply:
Indeed the fillDescriptions default is TRUE, but as any parameter it can be set to false.

Reply:
I just tested it on a sample of ephemeral 2023 PbPb data (HIEPhemeralHLTPhysics), and got the following results for the modified trigger ( HLT_HIUPC_ZeroBias_SinglePixelTrackLowPt_MaxPixelCluster400 ) :

Total events ran: 10 000
Events passing the trigger: 15
Uncompressed size of hltTriggerSummaryAOD : 7069 --> increased to --> 7122 bytes per event (+0.75%)
Compressed size of hltTriggerSummaryAOD : 509 --> increased to --> 518 bytes per event (+1.77%)

The trigger path added (518-509) * 10000 / 15 = 5.86 kB per fired event

Reply:
Regarding the additional parameter, how should I proceed? Do I leave it as it is currently (i.e. with the saveTags) or change the default value of saveTag in the class (so overwrite in the fillDescription) or add an additional parameter or move the for-loop that stores the candidates inside a if (saveTags) statement?

Reply:
> or change the default value of saveTag in the class (so overwrite in the fillDescription) 

In order to be more future-proof, I would be in favor of this (just for this filter) if the other agree. Otherwise I don't feel strongly provided we (collectively) keep memory of this change.

Reply:
Please do not change the default, that will just create confusion Rather, put saveTags to False (if you want False)  in the specific configured instances of this plugin in their python configuration.

Reply:
mmmh, the question is not about that. I think that they want this to be `True` in their own UPC trigger, but in general we want it to be false.

Reply:
Well, they are the only users.

Reply:
>  Well, they are the only users.

until now, yes. If someone tries to use it for pp we should remember about i.t

Reply:
> move the for-loop that stores the candidates inside an if (saveTags) statement?

I was also wondering if this (moving the call to `addObject` under an `if (saveTags())` statement) would be needed, and no, it is not needed as far as `triggerTriggerEvent_hltTriggerSummaryAOD__*` is concerned. This is shown in the example in [*] (the event size of "hlt1" is much smaller than that of "hlt0").

[*]
```bash
#!/bin/bash

# cmsrel CMSSW_14_0_11_MULTIARCHS
# cd CMSSW_14_0_11_MULTIARCHS/src
# cmsenv
# git cms-init --ssh
# git cms-remote add stahlleiton --ssh
# git cms-addpkg HLTrigger/special
# git checkout stahlleiton/HLT_UPC_CMSSW_14_1_X HLTrigger/special/plugins/HLTPixelTrackFilter.cc
# scram b

hltGetConfiguration /dev/CMSSW_14_0_0/GRun/V153 \
  --globaltag 140X_dataRun3_HLT_v3 \
  --data \
  --unprescale \
  --output minimal \
  --max-events 100 \
  --input /store/data/Run2024F/EphemeralHLTPhysics0/RAW/v1/000/382/250/00000/755afa10-e31c-41ba-afcd-8ab567a6951b.root \
  --paths MC_AK4PFJets_v27,HLTriggerFinalPath \
  > hlt0.py

cat <<@EOF >> hlt0.py

process.options.numberOfThreads = 1
process.options.numberOfStreams = 0

process.hltPixelTrackCands = cms.EDProducer( "ConcreteChargedCandidateProducer",
    src = cms.InputTag( "hltPixelTracks" ),
    particleType = cms.string( "pi+" )
)

process.hltOutputMinimal.outputCommands += [
    'keep trigger*_*_*_*',
]

process.hltPixelTrackCandsFilter = cms.EDFilter( "HLTPixelTrackFilter",
    saveTags = cms.bool( True ),
    pixelTracks = cms.InputTag( "hltPixelTrackCands" ),
    minPixelTracks = cms.uint32( 1 ),
    maxPixelTracks = cms.uint32( 9999999 ),
)

process.MC_AK4PFJets_v27.insert(-2, process.hltPixelTrackCands)
process.MC_AK4PFJets_v27.insert(-2, process.hltPixelTrackCandsFilter)

del process.hltAK4PFJetCollection20Filter
@EOF
cmsRun hlt0.py &> hlt0.log
mv output.root hlt0.root

cp hlt0.py hlt1.py
cat <<@EOF >> hlt1.py

process.hltPixelTrackCandsFilter.saveTags = False
@EOF
cmsRun hlt1.py &> hlt1.log
mv output.root hlt1.root

wget https://gist.githubusercontent.com/missirol/aed4d65ac12365fe27e72df3edc22193/raw/61d8aac29a02ce7cbbc8702c470de37df8db055f/hltFWLite_exa01.py -O printTriggerEvent.py
chmod u+x printTriggerEvent.py

./printTriggerEvent.py -v 10 -i hlt0.root > hlt0_triggerObjects.txt
./printTriggerEvent.py -v 10 -i hlt1.root > hlt1_triggerObjects.txt

edmEventSize -v hlt0.root > hlt0_eventSize.txt
edmEventSize -v hlt1.root > hlt1_eventSize.txt
```

```bash
> cat hlt0_eventSize.txt

File hlt0.root Events 100
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
triggerTriggerEvent_hltTriggerSummaryAOD__HLTX. 123333 61638.4
triggerTriggerEvent_hltTriggerSummaryAOD__HLT. 18126.6 5385.82
triggerTriggerEventWithRefs_hltTriggerSummaryRAW__HLTX. 67239.5 5229.64
triggerTriggerFilterObjectWithRefs_hltPixelTrackCandsFilter__HLTX. 67161.1 5195.37
l1tTauBXVector_hltGtStage2Digis_Tau_HLTX. 4326.87 946.17
l1tMuonBXVector_hltGtStage2Digis_Muon_HLTX. 1043.94 771.76
l1tEGammaBXVector_hltGtStage2Digis_EGamma_HLTX. 2534.24 763.47
l1tEtSumBXVector_hltGtStage2Digis_EtSum_HLTX. 5979.4 719.04
l1tJetBXVector_hltGtStage2Digis_Jet_HLTX. 1542.4 627.02
GlobalAlgBlkBXVector_hltGtStage2Digis__HLTX. 8294.99 512.84
GlobalAlgBlkBXVector_hltGtStage2ObjectMap__HLTX. 1948.03 378.41
GlobalExtBlkBXVector_hltGtStage2Digis__HLTX. 1483.27 138.52
edmTriggerResults_TriggerResults__HLT. 1767.08 111.36
EventAuxiliary 120.07 27.66
edmTriggerResults_TriggerResults__HLTX. 101.64 27.51
EventProductProvenance 248.54 12.49
EventSelections 79.21 4.66
BranchListIndexes 24.31 3.63

> cat hlt1_eventSize.txt

File hlt1.root Events 100
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
triggerTriggerEvent_hltTriggerSummaryAOD__HLT. 18126.6 5385.82
triggerTriggerEventWithRefs_hltTriggerSummaryRAW__HLTX. 67239.5 5229.64
triggerTriggerFilterObjectWithRefs_hltPixelTrackCandsFilter__HLTX. 67142.1 5169.63
l1tTauBXVector_hltGtStage2Digis_Tau_HLTX. 4326.87 946.17
l1tMuonBXVector_hltGtStage2Digis_Muon_HLTX. 1043.94 771.76
l1tEGammaBXVector_hltGtStage2Digis_EGamma_HLTX. 2534.24 763.47
l1tEtSumBXVector_hltGtStage2Digis_EtSum_HLTX. 5979.4 719.04
l1tJetBXVector_hltGtStage2Digis_Jet_HLTX. 1542.4 627.02
GlobalAlgBlkBXVector_hltGtStage2Digis__HLTX. 8294.99 512.84
GlobalAlgBlkBXVector_hltGtStage2ObjectMap__HLTX. 1948.03 378.41
GlobalExtBlkBXVector_hltGtStage2Digis__HLTX. 1483.27 138.52
edmTriggerResults_TriggerResults__HLT. 1767.08 111.36
triggerTriggerEvent_hltTriggerSummaryAOD__HLTX. 96.84 28.34
EventAuxiliary 120.07 27.64
edmTriggerResults_TriggerResults__HLTX. 101.64 27.51
EventProductProvenance 248.54 12.49
EventSelections 79.21 4.66
BranchListIndexes 24.31 3.63
```

Reply:
> mmmh, the question is not about that. I think that they want this to be True in their own UPC trigger, but in general we want it to be False.

I guess one could in principle do the following

(pseudo-code)
```diff
-void HLTFilter::makeHLTFilterDescription(edm::ParameterSetDescription& desc) { desc.add<bool>("saveTags", true); }
+void HLTFilter::makeHLTFilterDescription(edm::ParameterSetDescription& desc, bool const saveTags=true) { desc.add<bool>("saveTags", saveTags); }
```

then
```cpp
HLTPixelTrackFilter::fillDescriptions(  ....
[..]
  makeHLTFilterDescription(desc, false);
```
only in this specific plugin. Not sure how confusing/dangerous this is.

Reply:
> I guess one could in principle do the following

maybe I am missing something trivial, but I am not sure to see the need of this complication.
Can't we just do:

```diff
diff --git a/HLTrigger/special/plugins/HLTPixelTrackFilter.cc b/HLTrigger/special/plugins/HLTPixelTrackFilter.cc
index 5a374809479..b4b2d3c4fe0 100644
--- a/HLTrigger/special/plugins/HLTPixelTrackFilter.cc
+++ b/HLTrigger/special/plugins/HLTPixelTrackFilter.cc
@@ -52,7 +52,7 @@ HLTPixelTrackFilter::~HLTPixelTrackFilter() = default;
 
 void HLTPixelTrackFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
   edm::ParameterSetDescription desc;
-  makeHLTFilterDescription(desc);
+  desc.add<bool>("saveTags", false);
   desc.add<edm::InputTag>("pixelTracks", edm::InputTag("hltPixelTracks"));
   desc.add<unsigned int>("minPixelTracks", 0);
   desc.add<unsigned int>("maxPixelTracks", 0);
```

on top of this PR?
Using the script at https://github.com/cms-sw/cmssw/pull/45373#discussion_r1667067278 (in reverse order) seems to work just fine.

Reply:
```makeHLTFilterDescription(desc);``` should be there in case it is modified in the base class ```HLTFilter```.

Reply:
Honestly not sure what to make of the last comment.  I am not proposing at all to change the base class.

Reply:
If someone modifies the body of ```makeHLTFilterDescription(desc)``` https://cmssdt.cern.ch/lxr/source/HLTrigger/HLTcore/src/HLTFilter.cc#0025, then it will likely be forgotten to update/propagate it in this plugin where you dropped the call to ```makeHLTFilterDescription(desc)```.

Reply:
> it will likely be forgotten to update/propagate it in this plugin where you dropped the call to

Why is that an issue? It will then throw at runtime because of missing parameters, and immediately spotted.

Reply:
So let's avoid that by keeping this call as it is kept for all others.

Reply:
Mmmh, no. That generated the risk of having users to accidentally use this filter on heavy events and consequently accidentally increase substantially the event size, unless it is recalled when the corresponding configuration is integrated.

Reply:
Sorry, I just wholly disagree! In any case, if that is really an issue, and for me it is not,  we should definitely revisit ALL saveTags seetings of ALL instances and check with proponents if TRUE is really needed or can be flipped to false. There is no special case in this plugin.

Reply:
> In any case, if that is really an issue, and for me it is not, we should definitely revisit ALL saveTags seetings of ALL instances of all HLTFilters in the menu. 

I am not going to disagree to that.

> There is no special case in this plugin.

I think no other physics object in the menu can attain the multiplicity of pixel tracks, but maybe I am overestimating the impact on event size.